### PR TITLE
Removes ATOM/BOND_SPTR in boost::graph in favor of raw pointers

### PR DESCRIPTION
**Pull request by bp-kelley**
This removes the boost::shared_ptr usage in the boost graph.  It greatly reduces the memory footprint of a molecule and has a nice performance boost as well (at least on OS X).

Some more testing is required with valgrind to make sure it doesn't leak.

We may also consider making the lifetime of deleted atoms and bonds the lifespan of the molecule container to help prevent certain classes of issues when using the wrapped interfaces.

### CODE REVIEW

**Comment by greglandrum** [comment]
@bp-kelley : have you tested this with valgrind yet?

File: Code/GraphMol/RWMol.cpp [hunk 1]
```diff
@@ -267,6 +269,7 @@ void RWMol::removeAtom(Atom *atom) {
   boost::clear_vertex(vd, d_graph);
   // finally remove the vertex itself
   boost::remove_vertex(vd, d_graph);
+  delete atom;
```
Comment (greglandrum) [inline Code/GraphMol/RWMol.cpp]: I think this is going to result in leaking the memory for the bonds to the atom. The call to boost::clear_vertex() will remove the edges from the graph, but won't delete the bonds.

**Comment by greglandrum** [comment]
Though the performance improvements associated with this are really compelling, I'm concerned about there being loads of difficult to track down memory leak/corruption problems.

Could we get at least close to the same performance gain by using `std::unique_ptr` in the BGL data structures? That would make pointer ownership clear and would prevent the leaking problems.

**Comment by bp-kelley** [comment]
I tried using unique_ptr, but it has an issue with BGL because it can't be
copied.

I'll run some valgrind tests this week and see if there are any issues.

**Comment by bp-kelley** [comment]
btw - it could be that the performance increases are the result of not
deleting atoms/bonds so we won't really know until there aren't any leaks.

File: Code/GraphMol/RWMol.cpp (hunk 1 above)
Comment (bp-kelley) [reply Code/GraphMol/RWMol.cpp]: This is handled in the removeBonds function I think:

```
for (auto &nbr : nbrs) {
    removeBond(nbr.first, nbr.second);
  }
```

Comment (greglandrum) [reply Code/GraphMol/RWMol.cpp]: Sorry, I missed that. You're right.

**Comment by bp-kelley** [comment]
@greglandrum I've added the test-valgrind which appears to not leak in current testing.  The only thing I'm seeing is the periodic table leak which I think is fine.

File: Code/GraphMol/test-valgrind.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// $Id$
```
Comment (greglandrum) [inline Code/GraphMol/test-valgrind.cpp]: don't need svn droppings at the top of files anymore.

File: Code/GraphMol/test-valgrind.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+//
+//  Copyright (C) 2001-2008 Greg Landrum and Rational Discovery LLC
```
Comment (greglandrum) [inline Code/GraphMol/test-valgrind.cpp]: should probably fix this

File: Code/GraphMol/test-valgrind.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+// memory tests for valgrind
+void testRemoveAtomBond(RWMol &m, int atomidx, int bondidx)
+{
+  if (atomidx>=0) m.removeAtom(atomidx);
+  if (bondidx>=0) m.removeAtom(bondidx);
+}
+
+void testRemovals(RWMol m) {
+  boost::random::mt19937 rng;
```
Comment (greglandrum) [inline Code/GraphMol/test-valgrind.cpp]: using the RNG to pick which atoms to remove seems a bit heavy handed. How about just picking a few values?

File: Code/GraphMol/test-valgrind.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#include <boost/random/uniform_int.hpp>
+
+using namespace std;
+using namespace RDKit;
+
+// memory tests for valgrind
+void testRemoveAtomBond(RWMol &m, int atomidx, int bondidx)
+{
+  if (atomidx>=0) m.removeAtom(atomidx);
+  if (bondidx>=0) m.removeAtom(bondidx);
```
Comment (greglandrum) [inline Code/GraphMol/test-valgrind.cpp]: you mean `removeBond` here

**Comment by bp-kelley** [comment]
I'll fix these.  If we do accept this PR, we could also change the atom index to 32 bit from 16 as the molecules are quite a bit smaller.  This would supplant #1643

**Comment by greglandrum** [comment]
Looks like you checked in a test file (test-valgrind) that time

**Comment by greglandrum** [comment]
Switching to the 32bit atom index (in a different PR) makes sense.

**Comment by bp-kelley** [comment]
@greglandrum Yeah I added and then delete it.  Not enough sleep last night

**Comment by greglandrum** [comment]
Replaced by #1713

**Pull request closed without merging**